File: doi.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

type DOI struct {
	Author []struct {
		Family string
		Given  string
	}
	Issued struct {
		DateParts [1][3]*int `json:"date-parts"`
	}
	DOI            string
	Title          string
	Publisher      string
	Issue          *string
	Page           *string
	Volume         *string
	ContainerTitle *string `json:"container-title"`
	Published      *struct {
		DateParts [1][3]*int `json:"date-parts"`
	}
}

func (d *DOI) String() string {
	dateParts := d.Issued.DateParts
	if d.Published != nil {
		dateParts = d.Published.DateParts
	}
	date := ""
	if year := dateParts[0][0]; year != nil {
		date += fmt.Sprintf("%04d", *year)
		if month := dateParts[0][1]; month != nil {
			date += fmt.Sprintf("-%02d", *month)
			if day := dateParts[0][2]; day != nil {
				date += fmt.Sprintf("-%02d", *day)
			}
		}
	}

	authors := make([]string, 0, len(d.Author))
	for _, author := range d.Author {
		authors = append(authors, fmt.Sprintf("  - %s, %s", author.Family, author.Given))
	}

	page := ""
	if d.Page != nil {
		page = fmt.Sprintf(`
  page-range: "%s"`, *d.Page)
	}
	containerTitle := ""
	if d.ContainerTitle != nil {
		containerTitle = fmt.Sprintf(`
    title: "%s"`, *d.ContainerTitle)
	}
	volume := ""
	if d.Volume != nil {
		volume = fmt.Sprintf(`
    volume: "%s"`, *d.Volume)
	}
	issue := ""
	if d.Issue != nil {
		issue = fmt.Sprintf(`
    issue: "%s"`, *d.Issue)
	}

	return fmt.Sprintf(
		`  type: "article"
  title: "%s"
  author:
  %s
  date: "%s"%s
  serial-number:
    doi: "%s"
  parent:
    publisher: "%s"%s%s%s`,
		d.Title, strings.Join(authors, "\n  "), date, page, d.DOI, d.Publisher,
		containerTitle, volume, issue)
}

func getDoi(doi string) (*DOI, error) {
	resp, err := http.Get("https://citation.doi.org/metadata?doi=" + doi)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return nil, err
		}
		return nil, fmt.Errorf("HTTP status %d: %s", resp.StatusCode, body)
	}

	reference := new(DOI)
	if err := json.NewDecoder(resp.Body).Decode(reference); err != nil {
		return nil, err
	}

	return reference, nil
}

File: src/doi.rs
use std::fmt::{Display, Write};

use serde::Deserialize;

#[derive(Deserialize)]
struct Author {
    family: String,
    given: String,
}

#[derive(Deserialize, Clone, Copy)]
struct DateParts {
    #[serde(rename(deserialize = "date-parts"))]
    date_parts: ((Option<i64>, Option<i64>, Option<i64>),),
}

#[derive(Deserialize)]
pub struct Reference {
    author: Vec<Author>,
    issued: DateParts,
    #[serde(rename(deserialize = "DOI"))]
    doi: String,
    title: String,
    publisher: String,
    issue: Option<String>,
    page: Option<String>,
    volume: Option<String>,
    #[serde(rename(deserialize = "container-title"))]
    container_title: Option<String>,
    published: Option<DateParts>,
}

impl Display for Reference {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "  type: \"article\"")?;
        writeln!(f, "  title: \"{}\"", self.title)?;
        writeln!(f, "  author:")?;
        for author in &self.author {
            writeln!(f, "    - \"{}, {}\"", author.family, author.given)?;
        }

        let date_parts = self.published.unwrap_or(self.issued).date_parts.0;
        if let Some(year) = date_parts.0 {
            let mut date = format!("{year:04}");
            if let Some(month) = date_parts.1 {
                write!(date, "-{month:02}")?;
                if let Some(day) = date_parts.2 {
                    write!(date, "-{day:02}")?;
                }
            }
            writeln!(f, "  date: \"{date}\"")?;
        }

        if let Some(page) = &self.page {
            writeln!(f, "  page-range: \"{page}\"")?;
        }
        writeln!(f, "  serial-number:")?;
        writeln!(f, "    doi: \"{}\"", self.doi)?;
        writeln!(f, "  parent:")?;
        writeln!(f, "    publisher: \"{}\"", self.publisher)?;
        if let Some(container_title) = &self.container_title {
            writeln!(f, "    title: \"{container_title}\"")?;
        }
        if let Some(volume) = &self.volume {
            writeln!(f, "    volume: \"{volume}\"")?;
        }
        if let Some(issue) = &self.issue {
            write!(f, "    issue: \"{issue}\"")?;
        }

        Ok(())
    }
}
